### ISSUE TITLE
Add --date-from-filename to `osxphotos import`

### ISSUE THREAD

**Issue by RhetTbull** [RhetTbull/osxphotos#847]
A common problem when importing photos is setting the correct date (see #840). `osxphotos import` can set date from EXIF data but not from the filename.  #842 would allow a custom python function to do this but it would also be useful to be able to directly pull the date from the filename.  

Use case:

Imported files have the correct date [and time] embedded in the filename but EXIF data is not set or is not correct.  For example, consider the following filenames:

- `IMG1234_2022_11_17.jpeg`
- `IMG2134_20221117_0931.jpeg`

In the first example, the date is `2022-11-17` and in the second, the date/time is `2022-11-17T09:31:00`.  Many other variations are of course possible.  The python [strptime](https://docs.python.org/3/library/datetime.html#strftime-strptime-behavior) method of python's `datetime` can parse date strings using various format codes (see full list at [strftime.org](https://strftime.org/)).  `--date-from-filename` would use these patterns to parse the date [and time] from the filename.  For example:

`--date-from-filename "_%Y_%m_%d"` would match the first example above:

`_`: match underscore
`%Y`: match 4-digit year
`_`: match underscore
`%m`: match 2-digit month
`_`: match underscore
`%d`: match 2 digit day

Multiple formats could be provided separated by `|`:

`--date-from-filename "_%Y_%m_%d|_%Y%m%d_%H%M" `

would match either example above.

This would work by first attempting to match the first pattern then the second, and so on.

There would need to be a way to prevent false matches, for example, `%Y` prevented from matching the first 4 digits if filename was in format `IMG_1234_2022_11_17.jpeg`.  Possible solutions include a way to anchor the pattern, e.g. `_%Y_%m_%d.` to match the period in the filename or `%Y_%m_%d$` to anchor to end of file stem.  Other possibilities include full regex support but that would be harder to implement and mesh with the `strptime` format codes.

This could be implemented by creating a dynamically generated regex from the format codes which would help limit false matches.  Something like this:

`--date-from-filename "_%Y_%m_%d"` 

would become

`r"\_(\d{4})\_(\d{2})_(\d{2})"`

**Comment by RhetTbull** [comment]
Would be nice if `datetime.datetime.strptime()` could do this directly but it raises error if string doesn't match the pattern *exactly*.  However, `_strptime.py` from python source should be able to do this as it builds the regexes based on locale. I can use that code to build the dynamic regex:

https://github.com/python/cpython/blob/bbac9a8bcc4d7c0692e8b4f62a955f7ca107b496/Lib/_strptime.py#L170

**Comment by RhetTbull** [comment]
I've implemented my own `strptime` replacement at [strpdatetime](https://github.com/RhetTbull/strpdatetime) that will be a nice way to implement this feature.  For example, if you had filenames with dates embedded in different formats such as:

```
IMG_1234_2022_11_20.jpg
IMG_4567_2022_11_19.jpg
5678_20221101.jpg
```

You could parse these with:

`--date-from-file "^IMG_*_%Y_%m_%d|*_%Y%m%d"`

**Comment by RhetTbull** [comment]
Fixed in v0.54.4
